{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pybedtools\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TFBS Combinations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract GTEx Genexpression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "GTEx_with_tfbs = pybedtools.BedTool(\"/sybig/projects/GeneRegulation/data/jme/Bachelorarbeit/data/Promotor_with_TFBS/All_GTEx_Prom_with_TFBS.bed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "first = GTEx_with_tfbs[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['chr1',\n",
       " '17436',\n",
       " '17636',\n",
       " 'ENSG00000278267.1',\n",
       " '0',\n",
       " '-',\n",
       " 'MYCN',\n",
       " '17510',\n",
       " '17522',\n",
       " '-',\n",
       " 'miRNA',\n",
       " '0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,']"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "first.fields"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# with open(\"write_csv.csv\", \"w\") as f:\n",
    "#     for i in GTEx_with_tfbs[0:100]:\n",
    "#         f.write(f\"{i.name},{i.fields[11][:-1]}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"write_csv.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "test = pd.read_csv(\"/sybig/projects/GeneRegulation/data/jme/Bachelorarbeit/raw_data/gtex/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_median_tpm.gct\", sep=\"\\t\", skiprows=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Name,Description,Adipose - Subcutaneous,Adipose - Visceral (Omentum),Adrenal Gland,Artery - Aorta,Artery - Coronary,Artery - Tibial,Bladder,Brain - Amygdala,Brain - Anterior cingulate cortex (BA24),Brain - Caudate (basal ganglia),Brain - Cerebellar Hemisphere,Brain - Cerebellum,Brain - Cortex,Brain - Frontal Cortex (BA9),Brain - Hippocampus,Brain - Hypothalamus,Brain - Nucleus accumbens (basal ganglia),Brain - Putamen (basal ganglia),Brain - Spinal cord (cervical c-1),Brain - Substantia nigra,Breast - Mammary Tissue,Cells - Cultured fibroblasts,Cells - EBV-transformed lymphocytes,Cervix - Ectocervix,Cervix - Endocervix,Colon - Sigmoid,Colon - Transverse,Esophagus - Gastroesophageal Junction,Esophagus - Mucosa,Esophagus - Muscularis,Fallopian Tube,Heart - Atrial Appendage,Heart - Left Ventricle,Kidney - Cortex,Kidney - Medulla,Liver,Lung,Minor Salivary Gland,Muscle - Skeletal,Nerve - Tibial,Ovary,Pancreas,Pituitary,Prostate,Skin - Not Sun Exposed (Suprapubic),Skin - Sun Exposed (Lower leg),Small Intestine - Terminal Ileum,Spleen,Stomach,Testis,Thyroid,Uterus,Vagina,Whole Blood\n"
     ]
    }
   ],
   "source": [
    "csv_string = \",\".join(test.columns)\n",
    "print(csv_string)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filter for TFBS or TFBS combinations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = pybedtools.BedTool(\"/sybig/projects/GeneRegulation/data/jme/Bachelorarbeit/data/Promotor_with_TFBS/All_GTEx_Prom_with_TFBS.bed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\t17436\t17636\tENSG00000278267.1\t0\t-\tMYCN\t74\t86\t-\tmiRNA\t0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,\n",
      " chr1\t629439\t629639\tENSG00000225630.1\t730\t+\tOTX2\t-11\t1\t-\tunprocessed_pseudogene\t961.4,1024,949.35,328.1,497.15,363.6,518.8,1182,1160,1568,638.6,851.2,1275,990.9,1320,1213.5,1315,1591,1228,1440,902.9,207.95,180.7,370.8,427.8,826.4,777.5,814.1,335.2,798.7,510.5,1157,1283.5,1149,864.8,754.9,447.2,484.3,760.1,602.9,385.65,320.05,541.9,772.4,604.95,573.3,748.9,342.4,887.1,562,559,441.2,394.05,53.19,\n",
      " chr1\t634176\t634376\tENSG00000198744.5\t483\t+\tEGR2\t170\t181\t+\tunprocessed_pseudogene\t19.46,17.82,28.275,9.633,12.73,13.52,11.34,40.195,36.9,30.34,21.63,23.05,29.58,32.36,49.13,36.375,37.295,31.25,22.06,32.93,18.46,11.145,8.255,8.538,9.101,16.43,22.555,18.51,12.34,18.61,10.82,35.74,48.98,40.82,18.575,21.435,12.21,14.485,25.13,14,10.052,6.905,14.03,17.25,18.495,19.47,20.65,10.04,19.89,19.92,19.29,10.595,11.63,4.395,\n",
      " chr1\t758336\t758536\tENSG00000223181.1\t0\t-\tJUN,JUN,AR\t-4,-4,173\t12,8,190\t-,+,+\tsnRNA\t0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,\n",
      " chr1\t778569\t778769\tENSG00000237491.8\t262\t+\tREST,KLF9,SRF,SRF,USF2,KLF11,KLF15,KLF5,KLF1,E2F1,KLF4,KLF12,SP3,RXRA,USF1,JUN,JUND,TFE3,CREM,ATF7,ATF2,JUND,CREB1,NFE2L2,ATF3,JUN,PAX5,CREM,TCF7L2,TFAP2C,TFAP2C,PPARG,YY1,YY2,E2F1,NFE2L2,JUN,EGR1,TCF3,EGR2,BHLHE40,GRHL2,RFX3,PPARG,RFX5,RXRB,NEUROD1,MYB,MAFF,NFKB2,SP2,RELA,SP1,SP4,NFYC,THAP11,PKNOX1,RXRA,PBX2,EGR3,EGR1,EGR2,MLX,POU5F1,IRF3,NFYB,FOS,TP53,FOS,ELF1,NFYA,MYC,MYC,MNT,MAX,MAX,GLIS1,RUNX1,RUNX2,ZNF143,OSR2,HAND2,RUNX1,RELA,GABPA,SP1,TEAD3,THAP11,RELA,CTCF,CTCFL,JUND,CREM,NFIL3,ATF7,CREB1,CEBPA,ATF3,REL,RELB,NFKB1,RFX5,ESR1,ESR1,RXRA,JUN,JUND,ATF3,E2F1,E2F1,CREB1\t69,80,74,74,76,83,83,84,84,83,83,80,82,63,76,69,72,78,70,71,72,72,71,69,72,72,69,120,191,169,170,154,161,160,158,147,148,133,135,134,134,84,116,109,113,114,114,115,89,88,81,88,82,80,16,27,30,27,30,26,27,28,26,24,11,17,15,9,16,64,16,7,6,7,7,6,0,5,5,-8,-7,-6,-8,37,64,59,66,55,63,51,56,47,45,47,46,47,45,47,37,37,35,38,39,39,40,45,46,47,47,47,46\t90,96,90,90,90,94,94,94,95,95,95,95,95,80,90,88,87,86,86,85,85,85,84,84,84,84,81,136,205,184,182,174,173,171,170,162,161,147,146,145,144,96,130,129,129,128,127,125,105,101,98,98,97,97,27,46,45,44,43,41,41,39,36,35,32,29,28,27,27,78,27,18,18,17,17,17,16,16,14,8,5,4,4,47,78,74,74,74,73,70,68,62,61,60,60,60,60,59,47,48,48,54,56,56,57,58,59,59,59,59,59\t-,+,-,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,-,+,-,-,-,+,-,-,-,-,-,+,+,+,+,+,+,+,+,+,+,+,-,+,+,+,+,-,+,+,-,+,-,+,+,+,+,-,-,+,+,+,+,+,+,+,-,+,-,-,+,+,+,-,+,+,-,-,+,-,-,-,-,+,+,-,+,-,+,-,-,-,+,+,-,+,+,+,+,+,-,+,+,+,-,+,-,-,-,-,+,-\tlincRNA\t0.593,0.395,0.359,0.759,0.754,0.771,0.642,0.372,0.428,0.501,2.547,2.607,0.573,0.572,0.429,0.534,0.594,0.434,0.487,0.34,0.576,1.097,0.527,0.856,1.466,0.7,0.766,0.568,0.496,0.54,0.823,0.287,0.243,0.347,0.761,0.389,0.687,0.782,0.263,1.511,1.05,0.265,2.336,0.918,0.812,0.936,0.764,0.885,0.5,4.036,1.173,1.148,0.793,0.109,\n",
      " chr1\t778626\t778826\tENSG00000228327.3\t327\t-\tEGR1,GABPA,TFDP1,AR,ESR1,RXRA,REST,USF1,JUN,JUND,TFE3,CREM,ATF7,ATF2,JUND,CREB1,NFE2L2,ATF3,JUN,PAX5,SRF,GABPA,ELF1,SP1,TEAD3,THAP11,RELA,CTCF,CTCFL,JUND,CREM,NFIL3,ATF7,CREB1,CEBPA,RELA,E2F1,TCF3,EGR2,BHLHE40,CREM,RFX3,PPARG,RFX5,RXRB,NEUROD1,MYB,MAFF,NFKB2,SP2,E2F6,SP1,SP4,GRHL2,KLF9,SP3,KLF12,KLF4,E2F1,KLF1,KLF5,KLF15,KLF11,USF2,SRF,E2F1,MYC,MYC,MNT,MAX,MAX,GLIS1,RUNX1,RUNX2,ZNF143,OSR2,HAND2,RUNX1,STAT3,PPARG,ATF3,MYC,E2F1,MYC,MAX,MAX,NRF1,TP53,GABPA,AR,NR3C1,CTCF,CTCFL,USF2,CREB1,NFYC,NFYA,E2F1,E2F1,ATF3,JUND,JUN,RXRA,ESR1,ESR1,RFX5,NFKB1,RELB,RELA,REL,PKNOX1,FOS,TP53,FOS,NFYB,IRF3,POU5F1,MLX,EGR2,EGR1,EGR3,PBX2,RXRA,THAP11\t-4,192,192,190,190,63,53,53,55,56,57,57,58,58,58,59,59,59,59,62,53,65,65,69,69,69,70,73,75,81,82,83,83,83,83,45,190,-3,-2,-1,7,13,14,14,15,16,18,38,42,45,190,46,46,47,47,48,48,48,48,48,49,49,49,53,53,155,125,125,126,126,126,127,127,129,135,138,139,139,150,151,84,155,155,156,156,156,160,162,166,169,169,171,173,173,84,116,116,84,84,84,84,85,86,87,87,89,95,95,96,96,98,116,116,115,114,111,108,107,104,102,102,100,99,97\t10,206,203,207,207,80,74,67,74,71,65,73,72,71,71,72,74,71,71,74,69,79,79,84,77,88,80,92,87,96,98,96,97,96,98,55,202,8,9,9,23,27,34,30,29,29,28,54,55,62,203,61,63,59,63,61,63,60,60,59,59,60,60,67,69,167,136,137,136,136,137,143,138,138,151,150,149,151,161,166,96,167,167,169,166,166,171,180,180,186,186,193,185,187,97,127,127,96,96,96,97,98,103,104,104,105,108,106,106,106,113,127,134,128,126,132,119,117,115,116,117,113,116,116\t+,-,+,+,-,+,-,+,+,+,+,+,-,+,-,-,-,+,-,-,-,-,+,+,-,+,-,-,-,+,+,-,+,+,+,-,-,+,+,+,-,+,+,+,+,-,+,+,-,+,+,+,+,-,+,+,+,+,+,+,+,+,+,+,+,+,+,-,+,+,-,-,+,-,-,-,-,+,-,+,+,+,-,-,+,-,-,-,-,+,-,-,-,-,-,+,+,+,-,-,-,-,+,-,+,+,+,-,+,+,-,-,-,+,-,+,+,+,+,+,+,+,-,+\ttranscribed_unprocessed_pseudogene\t2.838,1.902,1.766,1.978,2.034,2.061,2.813,0.451,0.453,0.574,1.785,2.215,0.767,0.622,0.535,0.683,0.597,0.464,0.867,0.581,2.704,2.644,1.834,2.159,3.04,2.482,1.816,2.261,1.759,2.216,2.4,0.964,0.547,1.211,1.818,1.115,2.465,2.029,0.697,3.899,4.093,1.06,3.589,3.447,2.907,3.243,2.346,2.805,1.718,11.98,4.274,3.127,2.489,1.939,\n",
      " chr1\t817170\t817370\tENSG00000177757.2\t247\t+\tJUN,XBP1,RUNX1,RELA,FLI1,PPARG,SPI1,E2F1,CREB1,CREM,PRDM1,AR,CEBPB,SOX2,RARA\t-7,-3,22,25,28,24,16,48,48,46,63,71,81,114,30\t8,11,33,35,38,39,43,60,61,62,74,88,91,125,47\t-,+,-,-,+,+,+,-,+,-,+,+,+,+,+\tlincRNA\t1.92,0.871,1.25,1.252,0.98,1.871,0.834,0.232,0.219,0.241,1.266,1.635,0.239,0.227,0.211,0.333,0.278,0.22,0.286,0.214,0.951,0.513,0.927,0.89,0.833,0.615,0.512,0.668,0.321,0.49,0.959,0.369,0.195,0.264,0.191,0.138,0.566,0.595,0.245,1.243,0.784,0.149,0.725,0.715,0.723,0.955,0.496,0.315,0.389,1.527,0.897,0.819,0.721,0.075,\n",
      " chr1\t827522\t827722\tENSG00000225880.5\t386\t-\tESR1,KLF1,KLF1,SP1,SP1,SP2,KLF12,SP4,ARNT,HIF1A,MNT,BHLHE40,NRF1,MAX,HIF1A,AR,EGR3,EGR1,GLIS1,EGR2,E2F1,E2F1,MYC,MYC,NRF1,E2F6,MYCN,MYC,MYC,MAX,MAX,E2F1,KLF11,RXRA,NR2F6,NR2C2,RFX3,NR2F1,STAT5A,STAT3,NR2F2,NR2C1,RXRA,NR2F1,AR,FOXP1,FOXA3,FOXA2,SOX13,KLF15,SP3,SP1,SP2,KLF9,TFE3,MNT,MLX,KLF5,MYCN,E2F1,E2F6,CREM,FOXA1\t104,19,80,78,57,56,55,55,32,31,30,30,28,26,25,20,119,119,120,121,130,130,130,130,131,168,171,171,171,171,172,9,18,-16,-14,-14,-13,-13,-11,-10,-8,-7,-6,-6,-5,-5,-5,-5,-1,18,17,17,16,15,10,10,10,19,9,9,8,2,-5\t121,30,91,93,72,73,70,72,38,41,40,40,39,36,35,37,134,133,136,132,142,142,142,142,142,181,183,183,182,182,182,21,29,9,1,14,1,5,2,1,3,2,9,9,12,6,6,6,10,29,30,32,33,31,18,20,20,29,21,21,21,18,7\t+,-,-,-,-,-,-,-,+,+,+,+,-,-,-,-,+,+,+,+,+,-,+,-,-,-,-,-,+,+,-,-,-,+,+,+,+,+,-,-,+,+,-,-,+,+,+,+,+,-,-,-,-,-,-,-,-,-,+,+,-,-,+\tlincRNA\t4.49,3.425,2.752,6.226,5.935,7.817,7.573,1.454,1.423,2.043,11.1,10.96,2.051,1.739,1.866,1.99,2.046,1.525,3.332,1.85,4.099,3.921,3.334,9.637,10.475,5.755,4.147,4.791,3.824,4.378,5.094,2.752,1.464,3.572,5.979,2.851,6.654,4.872,1.817,11.39,12.04,0.979,9.092,12.84,4.155,4.325,4.028,5.879,2.775,7.369,9.106,9.471,9.455,1.251,\n",
      " chr1\t873091\t873291\tENSG00000234711.1\t129\t+\tCTCF,USF2,USF1\t108,80,80\t127,94,94\t-,+,+\tunprocessed_pseudogene\t0.058,0.063,0.072,0.057,0.06,0.057,0.099,0.135,0.156,0.129,0.115,0.125,0.186,0.159,0.127,0.157,0.156,0.117,0.092,0.079,0.091,0.051,0.038,0.064,0.074,0.08,0.106,0.066,0.078,0.062,0.083,0.124,0.145,0.101,0.033,0.066,0.075,0.049,0.114,0.058,0.068,0.037,0.223,0.09,0.243,0.172,0.155,0.077,0.08,0.376,0.073,0.062,0.123,0.035,\n",
      " chr1\t876754\t876954\tENSG00000230368.2\t129\t-\tCTCF\t-8\t11\t-\tlincRNA\t0.064,0.056,0.053,0.045,0.053,0.055,0.058,0.091,0.124,0.099,0.476,0.557,0.158,0.138,0.09,0.116,0.118,0.087,0.072,0.068,0.075,0.037,0.031,0,0.056,0.064,0.061,0.061,0.054,0.052,0.051,0.092,0.117,0.073,0.039,0.048,0.073,0.035,0.219,0.066,0.065,0.028,0.19,0.082,0.339,0.215,0.076,0.031,0.065,0.214,0.073,0.073,0.168,0.02,\n",
      " "
     ]
    }
   ],
   "source": [
    "data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " ### Genereating Subsets based on TFBS or TFBS pair in data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def func_for_tfbs_subset(BedTool_Interval, tfbs_lst):\n",
    "    \"\"\"\n",
    "    Function that returns TRUE, if all of the tfbs in tfbs_lst are in Interval.fields[6]\n",
    "    It also checks if tfbs in tfbs_lst occur more often then tfbs in Interval.fields[6]. \n",
    "    Therefore if tfbs_lst contains the same tfbs twice, it will only return True if Interval.fields[6] also contains the tfbs at least twice. \n",
    "    \"\"\"\n",
    "\n",
    "    # Processing tfbs_lst --> unique with counts\n",
    "    input_tfbs_unique, input_tfbs_counts = np.unique(np.array(tfbs_lst), return_counts=True)\n",
    "    input_tfbs_dict = dict(zip(input_tfbs_unique, input_tfbs_counts))\n",
    "    \n",
    "    # Processing tfbs in interval --> split tfbs, distance, strand into lists\n",
    "    tfbs_arr = np.array(BedTool_Interval.fields[6].split(\",\"))\n",
    "\n",
    "    # get unique tfbs and counts and save in dict\n",
    "    tfbs_unique, tfbs_counts = np.unique(tfbs_arr, return_counts=True)\n",
    "    tfbs_dict = dict(zip(tfbs_unique, tfbs_counts))\n",
    "\n",
    "    # Compare unique values and counts (equal or more (?))\n",
    "\n",
    "    # Checks if all unique TFBS in tfbs_lst occur in tfbs_unique.\n",
    "    if np.all(np.isin(input_tfbs_unique, tfbs_unique)):\n",
    "        \n",
    "        # Checks how often tfbs occur in input list, compared to interval tfbs. If tfbs count in input list is higher then in interval it will return False.\n",
    "        bool_lst = []\n",
    "        for tfbs, input_count in input_tfbs_dict.items():\n",
    "            tfbs_count = tfbs_dict[tfbs]\n",
    "            if input_count > tfbs_count:\n",
    "                bool_lst.append(False)\n",
    "            else:\n",
    "                bool_lst.append(True)\n",
    "        if np.all(bool_lst):\n",
    "            return True\n",
    "        else:\n",
    "            return False\n",
    "        \n",
    "        \n",
    "    else:\n",
    "        return False"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Generating subsets of tfbs_pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "tfbs_lst = [\"SMAD3\", \"ATF2\"]\n",
    "atf_smad_BedTool = data.filter(func_for_tfbs_subset, tfbs_lst).saveas()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\t1724311\t1724511\tENSG00000227775.3\t407\t+\tRXRB,NR2C2,RXRA,CTCFL,NR2C1,THRB,RXRA,YY1,NR2F2,HLF,HNF4G,HNF4A,RARA,NR2F1,PPARG,MYB,MAFF,CEBPA,ELF1,SPI1,JUNB,EHF,NR3C1,RELA,ELF3,REL,NFKB2,RELB,IRF2,NFKB1,NEUROD1,JUN,AR,TCF12,RUNX1,FOXO1,ESR1,ESR1,HAND2,NR2F6,ETV5,THAP11,RXRA,RUNX1,ELK1,ETS1,FLI1,ETV4,ETV6,ERG,FOXK1,ELK4,ETV1,GABPA,ELF3,SPI1,IRF4,PPARG,E2F1,RBPJ,CREB1,NFIL3,PDX1,SRF,MYC,NFE2L2,TBX5,MYC,NR2C1,NR2C2,ATF2,NR2F2,JUN,CTCF,JUN,CREM,FOXA2,FOXP1,FOXK2,FOXA1,REST,TCF7L2,HOXB13,MLX,MYC,POU2F2,POU5F1,NFE2,FOSL2,CREB1,TP53,ATF3,BATF,PBX2,ATF3,JUN,CEBPA,JUNB,FOS,JUND,MYC,BHLHE40,MYC,E2F1,ELF4,ELK1,ELK4,ELF1,ELF3,SPI1,GABPA,USF2,MYB,NRF1,REST,MAFK,MAX,AR,NR3C1,RARA,E2F1,EGR1,SP1,MYC,MYC,SP2,BCL6,NRF1,TFE3,FOSL2,ELF4,FLI1,ETS1,IRF4,ERG,ELK1,ERF,STAT1,ETV1,ARNT,FOSL1,SP1,SMAD3,YY1,MAFK,JUND,POU5F1,CREB1,JUN,ATF7,RXRA,NFE2L2,NR1H2,RFX5,RXRA,TFDP1,KLF5,E2F6\t134,133,129,137,140,134,131,138,139,149,137,134,131,134,133,144,142,148,89,86,66,88,87,94,90,94,91,94,87,92,135,95,87,113,116,119,117,117,137,133,175,161,166,173,175,175,175,175,175,175,166,175,173,173,174,170,176,193,197,190,162,150,159,155,161,158,165,161,165,166,150,164,163,157,151,161,167,167,167,167,57,60,61,65,52,63,66,62,66,65,53,66,67,65,66,63,65,67,66,66,15,-8,-7,-7,3,5,5,3,4,0,89,14,25,51,42,48,55,52,52,82,77,75,75,79,79,74,77,82,88,68,89,91,91,86,91,91,91,87,89,82,66,65,64,69,66,66,66,66,66,68,69,66,68,71,59,76,78,76\t148,148,149,149,149,149,150,150,150,163,150,150,151,152,153,154,158,160,103,106,79,103,104,104,104,104,104,105,105,105,148,108,118,124,127,130,134,134,147,148,186,180,181,184,185,185,185,185,185,185,180,186,187,187,188,190,191,213,209,200,175,163,167,171,172,173,173,173,174,174,163,175,175,176,176,177,178,178,178,179,78,74,75,75,76,76,77,77,78,78,71,78,78,78,78,78,79,79,79,79,27,2,4,5,15,15,16,17,18,20,103,28,35,62,63,63,65,69,69,99,89,89,90,91,91,91,93,93,96,79,101,101,101,101,101,101,101,102,103,88,79,80,80,81,81,81,81,82,82,82,84,84,85,87,87,87,88,89\t-,-,+,-,-,-,-,-,-,+,-,-,-,-,-,+,-,+,-,-,+,+,-,+,+,+,+,-,-,-,+,-,-,-,-,-,+,-,-,-,+,+,+,+,-,-,-,-,-,-,-,+,-,+,+,-,-,-,+,+,-,-,-,-,-,-,-,+,-,-,+,-,-,-,+,+,-,-,-,-,+,+,+,-,+,-,-,+,-,-,+,+,+,-,-,-,-,-,-,-,+,-,-,+,-,-,+,-,+,-,+,-,+,+,-,-,-,-,-,-,-,-,-,-,+,-,+,+,+,+,-,-,-,-,-,-,-,+,-,+,+,+,+,+,-,+,+,+,+,-,+,+,-,-,-,+,-,+\ttranscribed_processed_pseudogene\t6.996,4.232,5.829,6.674,5.982,5.087,8.894,2.561,3.714,2.846,14.08,17.02,5.909,5.01,2.792,2.866,3.605,2.039,3.188,1.922,7.48,7.701,6.531,12.65,17.745,8.281,7.341,6.856,4.544,6.37,10.35,2.589,1.73,3.447,3.242,1.964,7.348,7.512,2.538,9.263,15.14,3.314,12.19,10.04,6.533,7.265,9.169,9.992,5.995,4.314,12.69,14.995,9.06,3.062,\n",
      " chr1\t1724324\t1724524\tENSG00000008128.22\t478\t-\tTHRB,RXRA,ESR1,HAND2,NR2F6,NEUROD1,RXRB,NR2C2,RXRA,CTCFL,NR2C1,ESR1,YY1,NR2F2,HNF4G,HNF4A,RARA,NR2F1,PPARG,MYB,MAFF,RELB,GABPA,ETV1,ELF1,EHF,NR3C1,RELA,ELF3,REL,NFKB2,CEBPA,IRF2,NFKB1,SPI1,JUN,AR,TCF12,RUNX1,FOXO1,ETV4,FOXK2,FOXA1,FOXK1,THAP11,RXRA,RUNX1,ELK1,ETS1,FLI1,FOXP1,ETV6,ERG,ETV5,ELK4,ETV1,GABPA,ELF3,SPI1,IRF4,NR2C1,HLF,ATF2,NFIL3,PDX1,SRF,MYC,NFE2L2,TBX5,MYC,BHLHE40,NR2C2,CREB1,NR2F2,JUN,CTCF,JUN,CREM,FOXA2,CREB1,TP53,TCF7L2,HOXB13,MLX,MYC,POU2F2,POU5F1,NFE2,FOSL2,NR3C1,REST,ATF3,BATF,PBX2,ATF3,JUN,CEBPA,JUNB,JUNB,SPI1,STAT1,E2F1,MYC,E2F1,ELF4,ELK1,ELK4,ELF1,ELF3,ERF,MYC,USF2,MYB,NRF1,REST,MAFK,MAX,AR,NRF1,ARNT,E2F6,E2F1,EGR1,SP1,MYC,MYC,SP2,BCL6,JUND,TFE3,RARA,ELF4,FLI1,ETS1,IRF4,ERG,ELK1,FOS,TFDP1,FOSL1,FOSL2,SP1,SMAD3,YY1,MAFK,JUND,POU5F1,CREB1,JUN,ATF7,RXRA,NFE2L2,NR1H2,RFX5,RXRA,KLF5\t38,37,53,40,39,39,39,39,38,38,38,53,37,37,37,37,36,35,34,33,29,82,84,84,84,84,83,83,83,83,83,27,82,82,81,79,69,63,60,57,2,9,8,7,7,6,3,2,2,2,9,2,2,1,1,0,0,-1,-3,-4,13,24,24,24,20,16,15,14,14,14,185,13,12,12,12,11,11,10,9,109,116,113,112,112,111,111,110,110,109,118,109,109,109,109,109,109,108,108,108,167,85,194,183,182,172,172,171,170,169,86,160,159,152,125,124,124,122,118,94,99,98,98,98,97,96,96,96,94,108,91,88,86,86,86,86,86,86,108,100,108,108,107,107,106,106,106,106,105,105,105,103,103,102,100,100,99\t53,56,70,50,54,52,53,54,58,50,47,70,49,48,50,53,56,53,54,43,45,93,98,98,98,99,100,93,97,93,96,39,100,95,101,92,100,74,71,68,12,20,20,21,26,21,14,12,12,12,20,12,12,12,12,14,14,13,17,11,22,38,37,37,28,32,26,29,22,26,195,21,25,23,24,30,36,26,20,122,134,127,126,122,135,124,121,125,121,135,130,121,120,122,121,124,122,120,121,187,100,206,194,194,184,182,182,184,183,96,172,173,162,136,145,139,132,135,105,105,111,110,112,112,108,108,113,110,121,99,105,98,96,96,101,96,96,121,111,121,119,122,123,118,121,121,121,121,121,119,118,121,119,116,128,109\t-,-,-,-,-,+,-,-,+,-,-,+,-,-,-,-,-,-,-,+,-,-,+,-,-,+,-,+,+,+,+,+,-,-,-,-,-,-,-,-,-,-,-,-,+,+,+,-,-,-,-,-,-,+,+,-,+,+,-,-,-,+,+,-,-,-,-,-,-,+,-,-,-,-,-,-,+,+,-,-,+,+,+,-,+,-,-,+,-,-,+,+,+,-,-,-,-,-,+,-,+,-,-,+,-,-,+,-,+,-,+,-,+,+,-,-,-,-,+,+,+,-,-,-,-,+,-,+,-,+,-,-,-,-,-,-,-,-,+,+,+,+,+,+,-,+,+,+,+,-,+,+,-,-,-,-\tprotein_coding\t22.39,18.59,16.34,21.875,19.335,19.91,24.9,5.615,5.653,6.361,29.87,33.64,9.232,7.363,6.089,7.351,6.272,5.871,9.748,6.901,22.43,17.975,15.14,29.62,31.5,25.54,24.52,23.06,15.62,22.68,28.66,7.795,5.678,11.5,11.195,10.45,31.745,18.075,10.52,26.62,27.35,8.572,27.08,28.48,22.395,22.94,35.75,43.85,19.08,37.9,30.5,32.37,25.145,14.2,\n",
      " chr1\t19210276\t19210476\tENSG00000127481.14\t501\t-\tREST,E2F1,FOXK2,TCF7,USF1,AR,AR,BHLHE40,MLX,RUNX1,POU2F2,PAX5,SP1,ARNT,TFE3,ESR1,E2F4,KLF5,NR2C2,MYC,MAX,ATF7,YY1,REST,CREB1,NR3C1,RXRA,NR2F6,MITF,NFE2L2,NR2C2,NR2F1,RARA,MEIS2,PAX5,USF1,RFX5,NR2F2,NR2C1,NR2F1,NR2F6,RXRA,MYB,NR2F6,TCF7L2,SOX13,TCF7L1,REST,VDR,T,RARA,RFX3,RBPJ,RARA,NFE2L2,PAX5,GABPA,PAX5,ZNF143,E2F1,E2F1,E2F6,TFDP1,CEBPB,ZFX,MYCN,E2F1,HIF1A,MYC,SP2,MXI1,MXI1,MNT,MAX,E2F6,MYCN,KLF5,PPARG,FOXO1,NR4A1,HNF4A,RXRA,HNF4G,NR2F1,RXRB,SP3,EGR3,CTCFL,EGR1,EGR2,SP4,SP2,SPI1,IRF2,STAT1,SP1,USF2,KLF11,RXRA,RUNX1,ETV1,GABPA,ELF1,ELF3,NR1H2,FLI1,ERF,REL,CTCFL,CTCF,REST,TFAP2C,TFAP2A,TFAP2C,TP53,TFAP2A,RELB,NFKB1,E2F1,RELA,TFAP2C,TCF12,GLIS1,GLIS3,CTCF,T,GLIS2,RARA,CEBPA,CEBPD,ATF4,ATF7,BATF,NFIL3,NFE2,ATF3,CREB1,ATF3,E2F1,NFKB2,NEUROD1,PBX2,NFE2L2,HIF1A,ESR1,ESR1,TEAD3,E2F4,ETS1,CEBPG,ATF2,ELK1,ELK4,RXRA,ETV5,ERG,NFKB1,JUN,JUN,RELA,JUND,HIF1A,JUNB,CEBPA,CREB1,HLF,THAP11,SMAD3,CREM\t88,98,89,89,90,91,91,92,92,92,93,93,94,94,94,96,97,97,97,98,98,74,75,75,75,75,75,77,77,77,77,78,85,78,78,79,80,83,84,85,85,85,145,110,111,111,113,117,119,120,127,139,142,110,153,159,171,181,189,194,194,194,196,46,103,98,98,98,98,98,99,99,99,99,100,98,104,105,108,108,108,108,108,109,110,35,31,32,32,33,33,34,34,35,35,35,30,36,36,37,37,37,37,38,38,39,39,1,73,-11,-10,-9,-9,-8,-7,-7,-1,0,69,1,6,7,23,24,27,28,29,30,47,44,44,44,45,45,45,45,45,45,45,39,49,51,57,57,60,60,65,68,39,44,44,39,39,39,39,39,40,41,42,42,42,42,44,44,44,44,43,43,43\t109,110,100,100,104,124,108,102,102,103,106,105,109,100,102,113,111,107,105,110,109,88,87,96,88,108,100,92,95,92,92,96,103,86,90,93,96,94,93,100,100,100,155,125,125,122,125,155,134,136,145,153,152,127,168,171,185,193,205,206,206,215,207,56,117,110,110,106,110,115,109,109,109,109,113,110,114,125,119,120,124,125,121,124,124,48,46,44,46,44,50,51,54,53,50,50,44,47,54,48,51,51,51,52,55,49,49,11,85,8,11,5,5,5,11,5,10,13,81,11,19,18,39,38,46,44,43,47,59,57,58,58,56,58,56,57,58,57,57,52,62,64,72,67,77,77,73,82,49,59,57,49,50,54,50,49,53,68,60,52,57,52,57,58,57,58,62,57,59\t+,+,+,-,+,+,-,+,+,-,-,+,+,+,+,-,-,+,-,+,+,-,-,-,-,+,+,+,-,+,+,+,+,-,-,-,+,+,+,-,-,-,+,-,-,-,-,-,+,+,+,-,+,-,+,-,-,+,-,-,+,+,+,+,+,-,-,+,-,-,+,-,+,-,+,+,-,-,+,-,-,-,-,-,-,-,-,+,-,-,-,-,+,+,-,-,+,-,-,-,+,-,+,-,-,+,+,+,-,-,-,-,+,+,+,-,-,-,-,+,+,-,+,+,+,+,-,+,+,-,+,-,+,+,+,-,-,+,+,-,-,+,-,-,-,+,-,+,+,+,+,+,-,+,-,+,+,+,-,-,+,+,+,-,+,+,+,+,+\tprotein_coding\t29.92,22.52,24.395,43.235,33.89,39.77,47.21,7.752,9.915,10.315,27.57,33.79,17.25,13.12,8.186,10.68,12.115,7.74,10.72,8.374,26.84,49.77,36.555,34.49,37.69,35.13,22.265,43.8,21.97,42.87,31.19,12.38,9.961,16.99,17.36,12.305,36.405,22.69,30.74,38.94,45.705,10.325,33.18,31.11,34.805,41.23,27.58,25.06,22.81,66.91,35.93,45.075,29.47,10.8,\n",
      " chr1\t19210300\t19210500\tENSG00000230424.1\t287\t+\tMLX,RUNX1,POU2F2,PAX5,SP1,ARNT,TFE3,ESR1,E2F4,KLF5,NR2C2,MYC,MAX,E2F1,MYCN,MYCN,E2F1,HIF1A,MYC,SP2,NR2F1,NFE2L2,NR2C2,NR2F1,MEIS2,PAX5,USF1,RFX5,NR2F2,NR2C1,BHLHE40,NR2F6,RXRA,RARA,REST,FOXK2,TCF7,USF1,AR,AR,PAX5,VDR,T,RARA,RFX3,RBPJ,MYB,NFE2L2,PAX5,GABPA,REST,ZNF143,E2F1,E2F1,E2F6,TFDP1,E2F6,NRF1,MYC,MYC,HNF4A,MXI1,MNT,MAX,E2F6,ZFX,KLF5,PPARG,FOXO1,NR4A1,MXI1,RXRA,HNF4G,NR2F1,RXRB,RARA,NR2F6,TCF7L2,SOX13,TCF7L1,NFKB2,RXRA,RUNX1,ETV1,GABPA,ELF1,ELF3,NR1H2,FLI1,ERF,ETS1,KLF11,ELK1,ELK4,RXRA,ETV5,ERG,NFKB1,JUN,JUN,RELA,CTCFL,MITF,GLIS1,GLIS3,CTCF,T,GLIS2,RARA,USF2,EGR3,NR2F6,EGR1,EGR2,SP4,SP2,SPI1,IRF2,STAT1,SP1,SP3,E2F4,CEBPB,CEBPA,NEUROD1,PBX2,NFE2L2,HIF1A,ESR1,ESR1,TEAD3,JUND,E2F1,CTCFL,ATF7,YY1,REST,CREB1,NR3C1,RXRA,HIF1A,ATF3,E2F1,CREM,SMAD3,THAP11,HLF,CREB1,CEBPA,JUNB,CEBPG,CEBPD,ATF4,ATF7,BATF,NFIL3,NFE2,ATF3,CREB1,ATF2\t122,121,118,119,115,124,122,111,113,117,119,114,115,114,114,114,114,118,114,109,124,132,132,128,138,134,131,128,130,131,122,124,124,121,115,124,124,120,100,116,31,90,88,79,71,72,69,56,53,39,69,19,18,18,9,17,-6,-5,-8,-7,100,115,115,115,111,107,110,99,105,104,115,99,103,100,100,97,99,99,102,99,172,170,176,173,173,173,172,169,175,175,175,177,175,174,170,174,175,171,156,164,172,180,129,185,186,178,180,181,177,180,178,132,178,180,174,173,170,171,174,174,176,142,168,165,162,160,152,157,147,147,151,167,143,139,136,137,128,136,116,124,172,167,167,165,167,162,166,167,166,167,165,167,166,166,168,166,168,167,166,167\t132,132,131,131,130,130,130,128,127,127,127,126,126,126,126,126,126,126,126,126,139,147,147,146,146,146,145,144,141,140,132,139,139,139,136,135,135,134,133,133,43,105,104,97,85,82,79,71,65,53,107,35,30,30,30,28,7,6,4,4,116,125,125,125,124,121,120,119,116,116,125,116,116,115,114,114,114,113,113,111,185,188,187,187,187,187,186,186,185,185,185,188,185,185,185,185,185,184,183,182,182,192,147,201,200,197,196,195,194,194,193,147,192,191,191,190,190,189,189,189,189,156,178,177,175,173,167,167,164,164,159,182,155,151,150,149,149,149,149,149,182,179,179,181,181,181,180,180,180,180,180,180,180,180,179,179,179,179,179,180\t+,-,-,+,+,+,+,-,-,+,-,+,+,+,+,-,-,+,-,-,-,+,+,+,-,-,-,+,+,+,+,-,-,+,+,+,-,+,+,-,+,+,+,+,-,+,+,+,-,-,-,-,-,+,+,+,-,+,-,+,-,-,+,-,+,+,-,-,+,-,+,-,-,-,-,-,-,-,-,-,-,-,-,+,-,+,-,-,+,+,+,-,+,-,+,-,+,+,+,-,-,+,-,+,+,+,+,-,+,+,-,+,-,-,-,-,+,+,-,-,-,+,+,+,-,+,-,-,-,+,-,+,-,-,-,-,-,-,+,+,+,+,+,+,+,+,+,+,-,+,+,-,+,-,+,+,+,-,-,+\tantisense\t1.67,1.598,1.264,1.211,1.292,1.059,1.204,0.855,0.737,0.901,1.231,1.294,0.621,0.748,0.875,0.89,0.767,0.824,1.587,1.228,1.491,0.933,0.932,2.288,2.258,1.401,0.759,1.485,0.449,1.506,2.076,1.714,0.828,0.405,0.98,0.244,1.095,1.19,0.721,1.73,2.8,0.566,1.328,1.717,0.946,0.962,0.854,0.975,0.841,1.582,1.472,1.897,1.323,0.443,\n",
      " chr1\t36464437\t36464637\tENSG00000116898.11\t514\t-\tPAX5,MITF,POU2F2,E2F4,E2F1,TFDP1,MLX,MYC,USF2,E2F6,MYCN,MAX,MYC,MYCN,MYCN,RBPJ,E2F6,NFKB2,HNF4A,RXRA,RXRA,KLF1,KLF5,KLF15,KLF11,KLF4,SP3,SP2,SP4,MYC,SP1,TFAP2C,TFAP2C,NR3C1,MAX,MAX,MYC,MYC,PPARG,CTCF,MAX,MYCN,MNT,MYC,MNT,ESR1,E2F6,FOXA1,RUNX2,AR,RUNX1,TAL1,TFE3,ARNT,MXI1,MAX,BHLHE40,NRF1,BHLHE40,MAFF,JUNB,HLF,JUND,THAP11,CREM,SOX6,PDX1,SMAD3,MAFK,NFE2L2,JUND,RXRA,NR1H2,TCF7L2,BACH1,FOS,RFX5,FOSL2,PAX5,HIF1A,JUN,USF1,SREBF1,RXRA,USF2,MITF,ESR1,PPARG,TCF12,AR,MXI1,MAX,MAX,MYCN,MYC,MYC,CEBPA,CEBPB,NR2C2,ATF3,XBP1,CEBPG,BATF,NFE2,ATF3,JUN,MAFK,CEBPD,CEBPA,ATF4,NFIL3,CREB1,ATF2,FOSL1,ATF7,NR3C1,NFIL3,CREB1,E2F1\t42,43,43,38,39,38,47,45,45,46,46,46,46,46,76,38,38,34,24,24,24,18,18,17,17,17,16,15,14,76,16,164,164,136,135,135,135,134,120,117,87,86,77,76,47,73,73,68,64,63,62,61,49,47,47,47,47,47,47,-5,2,2,2,1,1,1,1,1,0,0,0,-3,-4,-4,1,-6,-6,-7,-9,-9,-10,-12,-12,-12,-12,-14,-14,-15,13,3,11,11,11,10,10,10,5,4,4,3,3,2,3,3,3,3,2,2,2,2,2,2,2,2,2,3,3,3,3\t54,61,56,52,51,49,57,58,59,59,58,57,58,58,88,48,56,47,39,39,40,29,28,28,28,29,29,32,31,98,31,177,178,153,146,145,146,146,135,136,98,98,87,98,57,90,86,80,73,80,73,73,57,53,57,57,57,58,57,11,15,16,15,20,17,11,9,15,15,15,15,12,13,10,16,15,10,13,3,1,15,2,16,3,2,4,3,5,24,20,21,23,21,24,22,22,17,14,12,15,17,17,14,14,15,18,17,15,16,16,15,15,15,15,16,20,16,16,15\t-,+,+,+,+,-,+,-,+,+,-,+,+,+,-,-,-,+,+,-,+,-,-,-,-,-,-,-,-,+,-,+,-,-,-,+,-,+,-,-,-,+,+,-,-,+,-,-,-,-,+,-,+,+,-,-,+,-,-,-,+,+,-,+,+,-,-,+,-,+,+,+,-,-,+,-,-,+,+,+,+,+,-,-,+,+,+,-,+,+,+,+,-,-,-,+,+,+,-,-,-,+,+,+,+,-,+,-,-,+,-,+,+,+,-,+,+,-,+\tprotein_coding\t38.71,43.77,34.455,32.84,35.945,34.39,37.79,18.745,21.28,21.515,25.35,21.88,21.03,25.44,17.7,22.695,21.36,19.04,21.88,20.33,37.55,60.82,55.655,35.07,33.945,39.8,29.245,35.67,37.45,37.46,32.18,28.06,31.155,27.81,33.84,31.965,37.45,28.9,59.15,30.91,28.16,15.42,35.96,36.9,30.68,30.67,25.2,30.39,32.46,121.7,36.41,35.235,32.75,11.31,\n",
      " chr1\t112619596\t112619796\tENSG00000116489.12\t553\t+\tELF1,TP63,TP53,GABPA,TP53,STAT4,ETV5,ELK4,GATA2,ERG,ETV4,FLI1,STAT5B,ELK1,RFX5,PBX3,ERF,TP63,ETV1,ELF3,RELA,NFKB1,NFKB1,KLF11,MAFF,ESR1,ESR1,CTCFL,PLAG1,MEIS2,MYB,PKNOX1,PBX2,MYC,NFIL3,ATF4,CEBPA,CEBPB,NFE2L2,MAFK,USF2,NR3C1,CEBPA,MYC,NFE2,AR,TCF7L2,ZFX,ZNF143,JUN,NFKB1,STAT1,KLF12,STAT3,SP1,KLF5,GRHL2,KLF1,ESR1,ETS1,RBPJ,PAX5,RUNX1,JUN,RUNX2,JUND,ATF3,EHF,TFAP2A,TFAP2C,TFAP2A,TFAP2C,RXRA,ETV1,ELF1,GABPA,TCF7L2,ERG,FLI1,BCL6,EBF3,TFAP2C,RUNX1,TFAP2A,EGR2,POU5F1,CEBPB,CEBPA,RXRA,JUN,BATF,FOS,FOSL1,SMAD3,TBX21,JUN,RXRA,AR,STAT3,ARNT,TFAP2C,GLIS2,NRF1,E2F1,RBPJ,E2F6,TFDP1,EGR1,EGR3,SP2,KLF5,NR2F2,ATF3,ATF3,E2F1,ATF2,CREB1,KLF4,E2F1,CTCFL,HIF1A,NRF1,E2F1,XBP1,CREM,ATF7,SP1,CREB1,JUND,JUN\t46,42,42,46,42,45,48,48,48,48,48,48,43,48,42,71,48,42,46,47,51,49,49,52,49,49,49,62,67,74,77,72,74,6,10,9,9,11,5,3,4,0,12,5,6,0,-4,-9,-14,89,35,45,41,45,39,42,40,41,31,48,33,21,20,13,20,13,13,127,144,144,143,143,136,128,128,128,153,130,130,123,126,126,128,126,85,156,172,170,172,175,179,179,178,178,184,178,171,183,194,85,126,124,85,84,87,84,86,84,83,82,89,93,93,93,93,92,92,121,121,119,121,118,115,93,91,92,91,93,90,93\t60,60,60,60,60,59,59,59,59,58,58,58,58,58,58,88,58,60,60,61,61,62,62,63,65,66,66,74,81,82,87,87,87,17,23,23,23,21,20,18,18,17,24,17,17,17,10,5,2,105,48,56,56,56,54,52,52,52,48,58,43,33,31,29,29,26,25,142,158,158,157,157,153,142,142,142,167,140,140,139,139,139,139,138,96,171,182,184,189,190,190,190,191,192,194,194,196,200,205,91,138,138,96,96,97,97,97,98,98,99,99,104,105,105,105,105,105,133,133,131,129,129,127,107,107,106,106,106,105,105\t-,+,+,+,-,-,+,+,+,-,-,-,-,-,+,+,-,-,-,+,+,+,-,-,+,+,-,-,-,+,+,+,-,+,-,-,+,-,+,+,+,+,-,-,+,-,-,+,+,-,-,-,-,-,-,-,+,-,-,-,-,-,+,+,-,+,-,+,+,-,-,+,+,-,-,+,+,-,-,-,-,+,+,+,+,+,-,-,+,-,+,-,+,+,+,+,-,-,+,-,-,-,+,+,-,-,-,+,+,+,+,+,-,+,-,-,-,-,+,+,-,+,-,+,-,+,-,+,-,+\tprotein_coding\t63.55,71.85,49.93,60.155,64.965,44.2,59.05,23.92,25.21,23.505,43.23,31.73,22.08,32.43,23.61,37.355,24.5,19.38,51.62,33.32,59.22,222.8,163.1,73.38,62.695,47.51,59.775,42.16,129.1,41.75,57.01,21.45,11.52,26.95,41.4,29.185,101.05,52.685,17.91,64.52,52.685,22.345,36.8,58.35,94.695,88.98,75.05,124.9,43.53,85.42,67.49,63.27,97.725,111.3,\n",
      " chr1\t148402315\t148402515\tENSG00000231551.8\t290\t+\tTHAP11,EGR2,GABPA,SPDEF,FLI1,ERG,ETS1,ELF3,ELF1,RUNX1,GABPA,MYB,RXRA,CTCF,MYB,ERG,PKNOX1,ESR1,JUN,CREM,CREB1,E2F1,ATF3,ATF2,HIF1A,RELA,POU5F1,JUN,JUN,FLI1,GABPA,EGR3,RARA,OTX2,GATA2,PAX6,RUNX1,CEBPA,NFIL3,CEBPG,CEBPB,WT1,KLF9,SP4,KLF12,SP2,KLF15,SP1,HIF1A,ESR1,SPI1,JUN,JUN,PBX2,CREM,PKNOX1,KLF5,KLF1,KLF11,E2F1,SMAD3\t37,116,43,48,65,65,65,62,63,66,63,88,81,80,100,45,32,30,28,25,26,27,27,26,24,12,4,-4,-8,45,117,114,198,191,191,178,181,173,171,172,172,159,156,155,157,154,158,155,122,117,114,125,121,125,123,125,158,157,158,157,123\t56,127,57,59,75,75,75,76,77,77,77,98,98,99,110,55,47,47,41,41,39,39,39,39,32,22,19,9,6,55,131,129,216,203,202,192,192,185,184,182,182,173,172,172,172,171,169,170,132,134,134,137,137,138,139,140,168,168,169,169,139\t-,+,-,-,+,+,+,-,+,-,-,-,+,+,+,+,-,+,+,+,-,-,+,-,-,+,+,+,-,+,+,+,+,+,-,+,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,+,-,+,-,-,-,-,-,-,-\ttranscribed_unprocessed_pseudogene\t0.746,0.573,0.364,1.421,1.486,1.198,1.158,1.388,1.308,1.693,1.784,2.227,1.317,1.272,1.241,1.552,1.746,1.076,1.435,1.308,1.205,0.879,0.784,1.47,1.933,1.142,0.911,0.971,0.689,0.809,1.565,0.601,0.155,0.578,1.022,0.377,1.1,0.97,0.079,1.477,1.129,0.399,3.567,2.107,0.658,0.62,1.341,2.335,0.492,4.977,1.374,0.989,1.172,1.137,\n",
      " chr1\t153001546\t153001746\tENSG00000163209.14\t672\t+\tRUNX1,GRHL2,FOXA1,TEAD4,TEAD1,GRHL2,GATA3,JUN,ATF2,ATF3,ESR1,HMBOX1,FOS,NR1H2,RUNX2,JUN,SMAD3,PBX2,FOSL1,FOS,JUN,FOSL2,JUND,JUNB\t177,11,16,80,80,83,97,144,142,143,146,159,159,152,184,159,157,159,158,158,155,160,158,158\t188,23,28,92,93,95,105,155,155,155,163,169,170,169,193,174,173,172,171,171,171,171,171,171\t-,-,+,-,-,+,-,+,-,+,+,+,-,+,+,+,-,+,-,+,-,+,-,-\tprotein_coding\t0.691,1.043,1.906,1.397,1.385,0.72,5.733,0.234,0.329,0.221,0.895,1.894,0.622,0.346,0.276,0.263,0.308,0.283,0.181,0.306,0.949,1.164,1.22,11.22,3.515,1.132,1.506,3.172,13400,4.384,4.373,0.707,0.435,0.904,0.072,1.703,1.207,5.802,0.498,0.711,2.692,1.426,2.029,2.015,1.121,0.967,1.65,2.041,1.969,2.102,1.274,2.616,4285.5,0.483,\n",
      " chr1\t156816862\t156817062\tENSG00000027869.11\t329\t-\tPKNOX1,JUN,JUN,MEIS1,NR3C1,NR2F2,NR4A1,NR2C1,EGR1,SPI1,NR3C1,MYB,E2F1,GABPA,ELF4,EGR2,EGR1,EGR3,BCL6,NR2F2,TP53,HIF1A,TAL1,ATF2,ATF7,CREB1,CEBPA,CREM,PBX2,NFE2L2,SMAD3,BACH2,ELF1,PBX3,GATA1,GATA2,GLIS1,RUNX1,GATA3,GATA2,REST,TCF12,TCF3,TAL1,GATA1,TCF4,MAX,E2F6,FOXO1,NFKB2,ERG,RUNX1\t80,81,81,83,84,124,124,125,151,170,172,42,-7,29,29,26,24,24,24,-6,-6,-7,74,-7,-8,-8,-9,-9,-11,-12,-13,-15,29,79,74,74,65,52,51,49,43,42,41,40,40,40,38,34,33,31,31,29\t95,91,95,90,101,135,136,134,165,190,202,52,5,43,41,37,38,39,40,5,12,3,86,6,6,5,3,7,2,3,3,4,43,96,85,85,81,63,59,60,64,53,52,58,60,53,48,47,44,44,41,40\t-,+,-,-,-,-,-,-,+,-,-,+,-,-,+,+,+,+,+,+,-,-,+,-,-,+,-,-,-,-,+,-,+,-,+,+,-,+,+,-,-,+,-,-,-,-,+,-,-,-,+,-\tprotein_coding\t1.517,4.878,0.822,1.312,1.76,0.441,1.811,0.232,0.169,0.321,0.138,0.215,0.274,0.225,0.265,0.398,0.211,0.377,1.089,0.503,1.755,11.935,0.373,1.608,4.003,2.34,2.516,0.988,1.315,0.914,2.859,1.796,0.764,1.103,2.377,1.08,7.214,0.848,0.273,1.297,1.05,0.506,2.738,3.478,0.548,0.507,5.564,13.85,1.835,0.589,2.78,3.224,2.155,16.16,\n",
      " chr1\t162790501\t162790701\tENSG00000132196.13\t411\t+\tCEBPA,CREB1,SMAD3,JUN,CEBPG,CEBPB,SREBF1,SP2,SP1,SP4,KLF11,KLF5,KLF15,GATA3,JUN,ATF7,CEBPD,ATF4,ATF2,CEBPG,ATF3,JUN,CEBPB,JUN,CEBPA,PBX2,HAND2,SNAI2,MYC,TCF4,ASCL1,PKNOX1,MYCN,ASCL1,MAX,MYC,MAX,TFAP4,MYB,NRF1,TEAD1,TEAD3,TFAP2A,TFAP2A,NFKB2,TFAP4\t119,17,118,116,121,121,114,105,106,104,107,108,107,97,17,16,17,16,17,15,17,14,18,120,122,173,173,167,168,167,167,171,168,169,168,168,169,169,169,167,137,139,130,128,124,169\t133,30,132,132,131,131,128,122,121,121,118,118,118,105,31,30,30,30,30,30,29,29,28,135,134,186,183,180,180,180,180,186,180,179,179,179,179,179,179,178,150,147,142,142,137,179\t+,+,-,-,-,-,+,+,+,+,+,+,+,-,+,-,-,+,+,+,-,-,+,+,-,-,+,+,+,+,+,+,-,-,-,-,+,-,+,+,-,-,+,-,+,+\tprotein_coding\t6.334,6.327,13.48,6.424,5.742,5.517,6.987,3.095,3.575,2.742,5.825,5.439,4.313,4.384,4.086,5.644,2.696,2.401,9.663,3.685,7.577,21.315,4.518,14.08,11.83,7.364,6.829,5.041,9.45,4.445,6.324,2.717,1.333,4.732,7.373,18.43,12.855,7.585,0.888,10.11,12.405,3.074,15.07,10.63,9.13,9.097,8.789,9.502,5.491,4.192,11.05,8.681,10.765,3.176,\n",
      " "
     ]
    }
   ],
   "source": [
    "atf_smad_BedTool.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filter for geneType or chromosome if necessary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_bed(BedTool, column_number, filter_content):\n",
    "    \"\"\"\n",
    "    Function that filters a bed file by a specific content in a specific column.\n",
    "    \"\"\"\n",
    "    filtered_BedTool = BedTool.filter(lambda x: x.fields[column_number] == filter_content)\n",
    "    return filtered_BedTool"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filter for GeneType:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\t923727\t923927\tENSG00000187634.11\t386\t+\tPBX2,GLIS3,ARNT,BHLHE40,MYC,HIF1A,MYC,KLF9,RELB,PAX5,SP1,CREB1,RXRA,RXRA,TFAP2C,TFAP2A,TFAP2A,FOXA2,RUNX2,EGR1,JUN,ATF3,REST,CREB1,FOXO1,NFIC,TFAP2A,MYB,TCF7L2,YY1,CTCFL,E2F1,NRF1,E2F1,TFAP2A,TEAD4,NFE2,JUN,JUND,SMAD3,NFE2L2,JUN,ESR1,ESR1,ERG,BACH1,MAX,MYCN,MYC,E2F1,MYC,KLF5,FLI1,TFAP2C,TFAP2A,TCF12,ETS1,NEUROD1,GABPA,ETV1,MYC,CTCF,CTCF,THAP11,TFAP4,REST,SP1,RUNX1,TFAP2A\t74,64,76,74,73,75,73,70,91,75,72,78,79,76,86,86,87,29,123,-2,0,2,-7,2,9,19,64,31,27,31,48,54,55,52,62,167,149,145,148,148,148,148,151,151,95,174,182,181,181,181,182,184,95,126,126,94,95,93,93,93,94,91,93,97,108,104,112,121,126\t87,78,82,84,84,85,85,86,102,87,87,91,94,95,99,99,99,40,132,12,13,14,14,15,20,30,76,41,41,43,60,66,66,74,76,179,160,160,161,162,163,164,168,168,105,189,192,193,193,193,193,194,105,138,140,105,105,106,107,107,109,110,112,116,118,125,127,132,138\t+,+,-,-,-,-,+,+,+,-,+,+,-,+,-,-,+,-,+,-,-,+,-,+,+,-,+,-,-,-,-,+,+,-,-,+,+,-,+,+,+,+,+,-,+,-,-,-,-,-,+,+,+,+,+,+,+,+,-,+,+,+,-,+,+,-,+,-,-\tprotein_coding\t0.476,2.931,3.785,5.717,3.104,7.932,19.06,0.533,0.466,2.218,0.426,0.624,0.832,0.502,0.906,3.641,1.063,1.393,2.443,1.184,1.704,0.085,1.628,6.077,9.004,4.419,4.035,5.286,1.32,5.469,24.38,6.885,0.453,7.537,16.455,1.632,3.584,0.702,0.282,5.292,1.8,0.734,33.27,15.92,0.531,0.794,4.91,17.48,3.423,6.864,5.414,11.29,6.696,0.046,\n",
      " chr1\t959309\t959509\tENSG00000188976.10\t547\t-\tTFAP2C,TFAP2A,TFAP2C,TFAP2A,AR,NFE2,E2F1,FOXP1,MYC,MYCN,MYC,TFE3,RELA,GLIS1,ESR1,JUND,RUNX2,PAX5,SPI1,SP2,SP1,SP4,KLF9,SP3,KLF11,KLF5,SP1,KLF1,KLF5,RUNX1,KLF15\t76,76,77,78,88,25,91,96,99,99,100,100,116,121,167,174,62,63,-18,-1,0,0,1,2,3,3,23,29,30,62,3\t90,90,90,90,105,36,103,107,111,111,111,108,126,137,184,187,71,75,2,16,15,17,17,15,14,13,38,40,40,73,14\t-,+,+,-,-,-,-,-,+,+,-,-,-,-,-,+,+,+,-,+,+,+,+,+,+,+,-,+,+,-,+\tprotein_coding\t52.86,48.19,37.17,55.495,51.9,62.56,71.87,18.345,27.56,25.095,82.57,77.12,40.79,45.14,22.64,28.535,26.835,21.24,27.8,20.96,58.94,94.205,95.02,67.47,78.605,54.33,45.825,55.96,76.51,54.7,75.63,29.43,23.84,29.39,44.965,27.06,55.21,49.15,59.3,69.76,64.865,28.81,56.97,63.03,83.105,90.32,49.57,56.15,42.38,98.67,66.79,77.39,71.255,14.99,\n",
      " "
     ]
    }
   ],
   "source": [
    "prot_coding = filter_bed(data, 10, \"protein_coding\")\n",
    "prot_coding.saveas().head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filter for chromosome:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr22\t10961338\t10961538\tENSG00000283047.1\t114\t-\tFOXA1,ESR1,GRHL2\t112,83,145\t124,100,157\t+,+,-\tunprocessed_pseudogene\t0.106,0.061,0.053,0.091,0.081,0.117,0.116,0.029,0.033,0.033,0.111,0.087,0.039,0.054,0.026,0.035,0.035,0.035,0.031,0.032,0.101,0.15,0.104,0.133,0.123,0.106,0.063,0.08,0.044,0.087,0.17,0.041,0.026,0.044,0.073,0,0.087,0.061,0.026,0.108,0.125,0.02,0.089,0.074,0.056,0.06,0.06,0.099,0.057,0.456,0.093,0.138,0.065,0,\n",
      " chr22\t11066217\t11066417\tENSG00000279973.2\t28\t+\tMYC,CTCF,MYC\t146,-9,132\t157,10,144\t-,+,+\tlincRNA\t0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0.51,0,0,0,0,\n",
      " chr22\t12602265\t12602465\tENSG00000283023.1\t108\t+\tYY1,E2F6,USF2,REST,ESR1,TEAD3,ETV5,ELF3,ETV1,ELF1,GABPA,ELF4,IRF2,SPI1\t198,15,44,48,96,158,175,173,174,174,174,176,172,171\t210,28,58,69,113,166,186,187,188,188,188,188,190,191\t-,-,-,-,+,+,-,-,+,+,-,+,+,+\tunprocessed_pseudogene\t0.089,0.066,0.047,0.11,0.079,0.098,0.104,0.033,0.027,0.042,0.077,0.06,0.04,0.046,0.032,0.042,0.039,0.031,0.081,0.041,0.086,0.154,0.042,0.14,0.053,0.117,0.058,0.106,0,0.094,0.068,0.043,0.026,0.036,0.045,0,0.064,0.049,0,0.123,0.138,0.039,0.16,0.066,0.055,0.055,0.062,0.073,0.054,0.332,0.092,0.137,0.081,0,\n",
      " "
     ]
    }
   ],
   "source": [
    "chr22_data = filter_bed(data,0,\"chr22\")\n",
    "chr22_data.saveas().head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract Important Data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Important informations that should be easily accesible:\n",
    "- GeneID\n",
    "- TFBSs with TSS distances and Template/non-Template \n",
    "- TFBS-count in Promotor (TFBS unique (?))\n",
    "\n",
    "Other informations that could be adressed:\n",
    "- Genomic locations of Promotor\n",
    "- (Score of Genexpression (?) only in GTEx file slightly interesting)\n",
    "- Overlapping TFBS from diffrent TFs --> More than 50% (?) --> only one TF can bind --> TF sets that competes for binding?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get Dataframe for filtered TFBS combinations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_information_for_singel_tfbs_set(BedTool_Interval):\n",
    "    \"\"\"\n",
    "    Function that is splitting the infomration about the TFBS, such as TSS-distance and strand oriantation and saving it in a Dataframe.\n",
    "    \"\"\"\n",
    "    # Splitting infomration into numpy arrays\n",
    "    tfbs_arr = np.array(BedTool_Interval.fields[6].split(\",\"))\n",
    "    tfbs_close_tss_arr = np.array(BedTool_Interval.fields[7].split(\",\"), dtype=\"int\")\n",
    "    tfbs_dist_tss_arr = np.array(BedTool_Interval.fields[8].split(\",\"), dtype=\"int\")\n",
    "\n",
    "    # Get information if tfbs is on same strand as gene or not (True, False)\n",
    "    tfbs_strand = np.array(BedTool_Interval.fields[9].split(\",\"))\n",
    "    tfbs_strand_orientation = np.array([\"nT\" if i==BedTool_Interval.fields[5] else \"T\" for i in tfbs_strand])\n",
    "\n",
    "    # Generate a numpy array, that contains the GeneID, necessary for additional functions\n",
    "    geneID = np.array([BedTool_Interval.fields[3] for i in range(len(tfbs_arr))])\n",
    "\n",
    "    return geneID, tfbs_arr, tfbs_close_tss_arr, tfbs_dist_tss_arr, tfbs_strand_orientation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['T', 'T', 'nT', 'T', 'T', 'T', 'T', 'T', 'T', 'nT', 'T', 'T', 'T',\n",
       "       'T', 'T', 'nT', 'T', 'nT', 'T', 'T', 'nT', 'nT', 'T', 'nT', 'nT',\n",
       "       'nT', 'nT', 'T', 'T', 'T', 'nT', 'T', 'T', 'T', 'T', 'T', 'nT',\n",
       "       'T', 'T', 'T', 'nT', 'nT', 'nT', 'nT', 'T', 'T', 'T', 'T', 'T',\n",
       "       'T', 'T', 'nT', 'T', 'nT', 'nT', 'T', 'T', 'T', 'nT', 'nT', 'T',\n",
       "       'T', 'T', 'T', 'T', 'T', 'T', 'nT', 'T', 'T', 'nT', 'T', 'T', 'T',\n",
       "       'nT', 'nT', 'T', 'T', 'T', 'T', 'nT', 'nT', 'nT', 'T', 'nT', 'T',\n",
       "       'T', 'nT', 'T', 'T', 'nT', 'nT', 'nT', 'T', 'T', 'T', 'T', 'T',\n",
       "       'T', 'T', 'nT', 'T', 'T', 'nT', 'T', 'T', 'nT', 'T', 'nT', 'T',\n",
       "       'nT', 'T', 'nT', 'nT', 'T', 'T', 'T', 'T', 'T', 'T', 'T', 'T', 'T',\n",
       "       'T', 'nT', 'T', 'nT', 'nT', 'nT', 'nT', 'T', 'T', 'T', 'T', 'T',\n",
       "       'T', 'T', 'nT', 'T', 'nT', 'nT', 'nT', 'nT', 'nT', 'T', 'nT', 'nT',\n",
       "       'nT', 'nT', 'T', 'nT', 'nT', 'T', 'T', 'T', 'nT', 'T', 'nT'],\n",
       "      dtype='<U2')"
      ]
     },
     "execution_count": 149,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "*_, strand = get_information_for_singel_tfbs_set(test_int)\n",
    "strand"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [],
   "source": [
    "def expand_tfbs_information_to_df(BedTool):\n",
    "    geneID_all = np.array([])\n",
    "    tfbs_all = np.array([])\n",
    "    tfbs_close_tss_all = np.array([])\n",
    "    tfbs_dist_tss_all = np.array([])\n",
    "    tfbs_orientation_all = np.array([])\n",
    "\n",
    "    for BedTool_Interval in BedTool:\n",
    "        geneID, tfbs_arr, tfbs_close_tss_arr, tfbs_dist_tss_arr, tfbs_orientation = get_information_for_singel_tfbs_set(BedTool_Interval)\n",
    "        geneID_all = np.append(geneID_all, geneID)\n",
    "        tfbs_all = np.append(tfbs_all, tfbs_arr)\n",
    "        tfbs_close_tss_all = np.append(tfbs_close_tss_all, tfbs_close_tss_arr)\n",
    "        tfbs_dist_tss_all = np.append(tfbs_dist_tss_all, tfbs_dist_tss_arr)\n",
    "        tfbs_orientation_all = np.append(tfbs_orientation_all, tfbs_orientation)\n",
    "        \n",
    "\n",
    "    return pd.DataFrame({\"geneID\":geneID_all, \"TFBS\": tfbs_all, \"Distance_to_TSS_close\":tfbs_close_tss_all, \"Distance_to_TSS_dist\":tfbs_dist_tss_all, \"Orientation\": tfbs_orientation_all})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_atf_smad = expand_tfbs_information_to_df(atf_smad_BedTool)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneID</th>\n",
       "      <th>TFBS</th>\n",
       "      <th>Distance_to_TSS_close</th>\n",
       "      <th>Distance_to_TSS_dist</th>\n",
       "      <th>Orientation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000227775.3</td>\n",
       "      <td>RXRB</td>\n",
       "      <td>134.0</td>\n",
       "      <td>148.0</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000227775.3</td>\n",
       "      <td>NR2C2</td>\n",
       "      <td>133.0</td>\n",
       "      <td>148.0</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000227775.3</td>\n",
       "      <td>RXRA</td>\n",
       "      <td>129.0</td>\n",
       "      <td>149.0</td>\n",
       "      <td>nT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              geneID   TFBS  Distance_to_TSS_close  Distance_to_TSS_dist  \\\n",
       "0  ENSG00000227775.3   RXRB                  134.0                 148.0   \n",
       "1  ENSG00000227775.3  NR2C2                  133.0                 148.0   \n",
       "2  ENSG00000227775.3   RXRA                  129.0                 149.0   \n",
       "\n",
       "  Orientation  \n",
       "0           T  \n",
       "1           T  \n",
       "2          nT  "
      ]
     },
     "execution_count": 154,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_atf_smad.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
